File: frontend/lib/features/notifications/data/models/notification_model.dart
// lib/features/notifications/data/models/notification_model.dart
import '../../domain/entities/notification.dart';

class NotificationModel extends NotificationEntity {
  const NotificationModel({
    required super.id,
    required super.title,
    required super.body,
    required super.type,
    required super.priority,
    super.isRead = false,
    super.imageUrl,
    super.actionUrl,
    required super.createdAt,
    required super.ownerId,
  });

  factory NotificationModel.fromJson(Map<String, dynamic> json) {
    return NotificationModel(
      id:        json['id']       as String? ?? '',
      title:     json['title']    as String? ?? '',
      body:      json['body']     as String? ?? '',
      type:      json['type']     as String? ?? 'system',
      priority:  json['priority'] as String? ?? 'low',
      isRead:    json['isRead']   as bool?   ?? false,
      imageUrl:  json['imageUrl'] as String?,
      actionUrl: json['actionUrl'] as String?,
      ownerId:   json['ownerId']  as String? ?? '',
      createdAt: _parseDate(json['createdAt']),
    );
  }

  static DateTime _parseDate(dynamic v) {
    if (v == null) return DateTime.now();
    if (v is String) {
      // Handle both ISO strings and Firestore Timestamp maps
      return DateTime.tryParse(v) ?? DateTime.now();
    }
    // Firestore Timestamp object (from Dart SDK) has a toDate() but we
    // receive it as a map when deserialised from JSON: {_seconds, _nanoseconds}
    if (v is Map) {
      final secs = (v['_seconds'] as num?)?.toInt() ?? 0;
      return DateTime.fromMillisecondsSinceEpoch(secs * 1000);
    }
    return DateTime.now();
  }

  Map<String, dynamic> toJson() => {
        'id':        id,
        'title':     title,
        'body':      body,
        'type':      type,
        'priority':  priority,
        'isRead':    isRead,
        'imageUrl':  imageUrl,
        'actionUrl': actionUrl,
        'createdAt': createdAt.toIso8601String(),
        'ownerId':   ownerId,
      };
}

File: backend/models/notification_model.py
# backend/models/notification_model.py
from dataclasses import dataclass
from typing import Optional
from utils.helpers import ts_to_iso

@dataclass
class NotificationModel:
    id:         str
    owner_id:   str
    title:      str
    body:       str
    type:       str      # price_update | weather_alert | order | system
    priority:   str      # high | medium | low
    is_read:    bool
    created_at: str
    image_url:  Optional[str] = None
    action_url: Optional[str] = None

    def to_dict(self) -> dict:
        return {
            "id":        self.id,
            "ownerId":   self.owner_id,
            "title":     self.title,
            "body":      self.body,
            "type":      self.type,
            "priority":  self.priority,
            "isRead":    self.is_read,
            "imageUrl":  self.image_url,
            "actionUrl": self.action_url,
            "createdAt": self.created_at,
        }

    @classmethod
    def from_firestore(cls, doc_id: str, data: dict) -> "NotificationModel":
        return cls(
            id=         doc_id,
            owner_id=   data.get("ownerId",  ""),
            title=      data.get("title",    ""),
            body=       data.get("body",     ""),
            type=       data.get("type",     "system"),
            priority=   data.get("priority", "low"),
            is_read=    data.get("isRead",   False),
            created_at= ts_to_iso(data.get("createdAt")),
            image_url=  data.get("imageUrl"),
            action_url= data.get("actionUrl"),
        )
